src/GovUk.Education.ExploreEducationStatistics.Public.Data.Api/Services/DataSetService.cs

```diff
@@ -61,28 +70,40 @@ public DataSetService(PublicDataDbContext publicDataDbContext)
         string dataSetVersion)
     {
         return await CheckVersionExists(dataSetId, dataSetVersion)
+            .OnSuccessDo(_userService.CheckCanViewDataSetVersion)
             .OnSuccess(MapDataSetVersion);
     }
 
     public async Task<Either<ActionResult, DataSetVersionPaginatedListViewModel>> ListVersions(
+        Guid dataSetId,
         int page,
-        int pageSize,
-        Guid dataSetId)
+        int pageSize)
+    {
+        return await _publicDataDbContext.DataSets
+            .SingleOrNotFound(ds => ds.Id == dataSetId)
+            .OnSuccessDo(_userService.CheckCanViewDataSet)
+            .OnSuccess(dataSet => ListPaginatedVersions(dataSet: dataSet, page: page, pageSize: pageSize));
+    }
+
+    private async Task<Either<ActionResult, DataSetVersionPaginatedListViewModel>> ListPaginatedVersions(
+        DataSet dataSet,
+        int page,
+        int pageSize)
     {
         var queryable = _publicDataDbContext.DataSetVersions
-            .Where(ds => ds.DataSetId == dataSetId)
+            .Where(ds => ds.DataSetId == dataSet.Id)
             .Where(ds => ds.Status == DataSetVersionStatus.Published
-                || ds.Status == DataSetVersionStatus.Unpublished
-                || ds.Status == DataSetVersionStatus.Deprecated);
+                         || ds.Status == DataSetVersionStatus.Withdrawn
+                         || ds.Status == DataSetVersionStatus.Deprecated);
```

Comment:
I've been looking at this for a while wondering if we could possibly re-use the authorization logic we have in the authorization handlers here... Otherwise if we change the rule in one place we have to remember to change it here too...

I can't think of a way to do this currently, but It would be nice not to have this logic in 2 places. The authorization handlers currently work fine if checking against individual records, but isn't so great when looking for a list of records - we can't use it here unless we bring them all into memory first and then filter them :/

Have you got any ideas?

Reply:
As discussed on Slack, there isn't a great way to do this currently that isn't more involved due to how LINQ works. There's potentially ways to make this work better by wrangling with expressions, but in the interest of time, we'll continue with the current style.

To neaten things up some more, I've refactored these checks into more re-usable auth extension methods.